core/src/main/java/feign/DelegatingClient.java

```diff
@@ -0,0 +1,62 @@
+/*
+ * Copyright 2012-2022 The Feign Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package feign;
+
+import java.io.IOException;
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+
+/**
+ * A class that combines both sync and async clients and is capable of executing requests for both
+ */
+public abstract class DelegatingClient<C> implements Client, AsyncClient<C> {
```

Comment:
I see much effort is been spent on trying to solve the "what if" scenario a client implements both interfaces....

I think for v1 we shouldn't worry about that at all.

First, cause the method invoked wont be the same.  So, I think we could change `Capability#invoke` to call all matching methods, not just the first. 

What do you think?

Reply:
But we do have such clients already. `Http2Client` implements both. `OkHttpClient` does both as well. That's why I'm trying this... If I change `invoke` to call all methods, that will result in `new MeteredClient(new MeteredClient(okHttpClient))` because both `enrich(Client)` and `enrich(AsyncClient)` will do the same thing, and that won't be expected I think.

We can get away without this `DelegatingClient`. I extracted it because I need the same logic for all three metered clients with no other commonalities other than `feign-core`.